src/include/server/mir/shell/surface_ready_registrar.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright © 2015 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_SHELL_SURFACE_READY_REGISTRAR_H_
+#define MIR_SHELL_SURFACE_READY_REGISTRAR_H_
+
+#include <functional>
+#include <memory>
+#include <map>
+#include <mutex>
+
+namespace mir
+{
+namespace scene { class Session; class Surface; }
+namespace geometry { struct Size; }
+
+namespace shell
+{
+/// Notifies the user when a surface is ready
```

Comment:
I feel this could be better documented, or maybe could be easier to use? Particularly: how easy would it be for this to *not* have the `register_surface`/`unregister_surface` pair, and just fire the `ActivateFunction` when *any* surface becomes ready?